src/driver/connection.ts

```diff
@@ -149,32 +243,100 @@ export class Connection extends MongooseConnection {
         return this;
     }
 
-    async createClient(uri: string, options?: ConnectOptionsInternal) {
+    /**
+     * Create an astra-db-ts client and corresponding objects: client, db, admin.
+     * @param uri the connection string
+     * @param options
+     */
+    async createClient(uri: string, options: ConnectOptionsInternal) {
         this._connectionString = uri;
         this._closeCalled = false;
         this.readyState = STATES.connecting;
 
-        const client = await Client.connect(uri, options);
+        const { baseUrl, keyspaceName, applicationToken, baseApiPath } = parseUri(uri);
+
+        const dbOptions = { dataApiPath: baseApiPath };
+
+        const isAstra = options?.isAstra;
+        const { adminToken, environment } = isAstra
+            ? { adminToken: applicationToken, environment: 'astra' as const }
+            : {
+                adminToken: new UsernamePasswordTokenProvider(
+                    options.username || throwMissingUsernamePassword(),
+                    options.password || throwMissingUsernamePassword()
+                ),
+                environment: 'dse' as const
+            };
+        const client = new DataAPIClient(adminToken, { environment });
+        const db = options?.useTables
+            ? new TablesDb(client.db(baseUrl, dbOptions), keyspaceName)
+            : new CollectionsDb(client.db(baseUrl, dbOptions), keyspaceName);
+        const admin = isAstra
+            ? db.astraDb.admin({ adminToken })
+            : db.astraDb.admin({ adminToken, environment: 'dse' });
+
+        const collections: Collection[] = Object.values(this.collections);
+        for (const collection of collections) {
+            collection._collection = undefined;
+        }
+
         this.client = client;
-        this.db = client.db();
+        this.db = db;
+        this.admin = admin;
+        this.baseUrl = baseUrl;
+        this.keyspaceName = keyspaceName;
+        this.baseApiPath = baseApiPath;
 
         this.readyState = STATES.connected;
         this.onOpen();
         return this;
+
+        function throwMissingUsernamePassword(): string {
+            throw new AstraMongooseError(
+                'Username and password are required when connecting to self-hosted DSE',
```

Comment:
```suggestion
                'Username and password are required when connecting to non-Astra deployments',
```
Or something along these lines. The idea is to keep the wording open/inclusive (ideally throughout the library), especially since HCD is a thing now, succeeding DSE.